File: Source code/MATLAB/Simplex.m
function Simplex()
    % Ввод размерности
    N = input('Введите размер задачи N: ');

    % Инициализация таблицы стоимости
    cost = zeros(N, N);
    disp('Введите элементы таблицы стоимости:');
    for i = 1:N
        row = input('', 's'); % Чтение строки в виде строки символов
        cost(i, :) = str2num(row); % Преобразование строки в массив чисел
    end

    solveAssignmentProblem(cost, N);
end

% Функция для вывода текущей симплекс-таблицы
function printTable(table, basis, N)
    disp('Симплекс-таблица:');
    fprintf('\t');
    for i = 1:N*N
        fprintf('x%d\t', i);
    end
    fprintf('\n');
    fprintf('z\t');
    p = 1;
    for i = 1:size(table, 1)
        fprintf('%f\t', table(i, :));
        fprintf('\n');
        if p <= length(basis) && basis(p) ~= -1
            fprintf('x%d\t', basis(p) + 1);
        else
            fprintf('\t');
        end
        p = p + 1;
    end
    fprintf('Базисные переменные: ');
    for i = 1:length(basis)
        if basis(i) ~= -1
            fprintf('x%d ', basis(i) + 1);
        end
    end
    fprintf('\n\n');
end

% Функция для нахождения разрешающего столбца
function pivotCol = findPivotColumn(table, basis, N)
    pivotCol = -1;
    maxdelta = 0;
    for p = 1:N*N
        delta = -table(1, p);
        for i = 1:length(basis)
            delta = delta + table(1, basis(i) + 1) * table(i + 1, p);
        end
        if delta > maxdelta
            maxdelta = delta;
            pivotCol = p;
        end
    end
end

% Функция для нахождения разрешающей строки
function pivotRow = findPivotRow(table, pivotCol, N)
    INF = 1e9;
    pivotRow = -1;
    minRatio = INF;
    for i = 2:2*N + 1
        if table(i, pivotCol) > 0
            ratio = table(i, end) / table(i, pivotCol);
            if ratio < minRatio
                minRatio = ratio;
                pivotRow = i;
            end
        end
    end
end

% Функция для выполнения симплекс-поворота
function [table,basis] = pivot(table, basis, pivotRow, pivotCol)
    pivotValue = table(pivotRow, pivotCol);

    % Делим строку разрешающего элемента на сам элемент
    table(pivotRow, :) = table(pivotRow, :) / pivotValue;

    % Обновляем остальные строки
    for i = 2:size(table, 1)
        if i ~= pivotRow
            factor = table(i, pivotCol);
            table(i, :) = table(i, :) - factor * table(pivotRow, :);
        end
    end

    % Обновляем базисные переменные
    basis(pivotRow-1) = pivotCol - 1;
end

% Основная функция для решения задачи о назначениях с использованием симплекс-метода
function solveAssignmentProblem(cost, N)
    % Инициализация симплекс-таблицы и базисных переменных
    table = zeros(2 * N + 1, N * N + 1);
    basis = -ones(1, 2 * N);  % Обратите внимание на правильную инициализацию размера массива

    % Заполняем строку целевой функции
    for j = 1:N*N
        table(1, j) = cost(floor((j - 1) / N) + 1, mod(j - 1, N) + 1);
    end

    % Заполняем таблицу ограничений
    for i = 2:N + 1
        for j = 1:N
            table(i, (i - 2) * N + j) = 1;
        end
        table(i, end) = 1;
    end
    for i = N + 2:2 * N + 1
        for j = 1:N*N
            table(i, j) = mod(j - 1, N) == mod(i - N - 2, N);
        end
        table(i, end) = 1;
    end

    % Заполнение массива basis
    for i = 1:N
        basis(i) = (i - 1) * N;
    end
    currentIndex = N + 1;
    for i = 1:N*N
        if isempty(find(basis(1:N) == i - 1, 1))
            basis(currentIndex) = i - 1;
            currentIndex = currentIndex + 1;
        end
        if currentIndex > 2 * N
            break;
        end
    end

    disp('Начальный базис построен');
    printTable(table, basis, N);

    % Симплекс-метод
    iteration = 1;
    while true
        pivotCol = findPivotColumn(table, basis, N);
        fprintf('Ведущий столбец: %f\n', pivotCol);
        if pivotCol < 0
            break;
        end
        pivotRow = findPivotRow(table, pivotCol, N);
        fprintf('Ведущая строка: %f\n', pivotCol);
        if pivotRow == -1
            disp('Решение не существует (неограниченная задача).');
            return;
        end

        [table, basis] = pivot(table, basis, pivotRow, pivotCol);
        fprintf('Итерация %d\n', iteration);
        iteration = iteration + 1;
        printTable(table, basis, N);
    end

    % Вывод оптимального решения
    totalSum = 0;
    disp('Оптимальное решение найдено:');
    for i = 1:length(basis)
        if table(i + 1, end) ~= 0
            col = mod(basis(i), N) + 1;
            row = floor(basis(i) / N) + 1;
            totalSum = totalSum + cost(row, col);
            fprintf('Работник %d назначен на задачу %d\n', row, col);
        end
    end
    fprintf('Сумма значений, соответствующих номерам в базисе: %f\n', totalSum);
end

File: Source code/C++/Simplex.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits>

const int INF = std::numeric_limits<int>::max();
int N; // Глобальная переменная N для размерности задачи

// Функция для вывода текущей симплекс-таблицы
void printTable(const std::vector<std::vector<double>>& table, const std::vector<int>& basis) {
	std::cout << "Симплекс-таблица:" << std::endl;
	std::cout << "\t";
	for (int i = 0; i < N * N; ++i) {
		std::cout << 'x' << i + 1 << "\t";
	}
	std::cout << std::endl << 'z' << "\t";
	int p = 0;
	for (const auto& row : table) {
		for (double value : row) {
			std::cout << value << "\t";
		}
		std::cout << std::endl;
		if (basis[p] != -1) {
			std::cout << 'x' << basis[p] + 1 << "\t";
		} else {
			std::cout << "\t";
		}
		p++;
	}
	std::cout << "Базисные переменные: ";
	for (int i = 0; i < basis.size(); ++i) {
		if (basis[i] != -1) {
			std::cout << "x" << basis[i] + 1 << " ";
		}
	}
	std::cout << std::endl << std::endl;
}

// Функция для нахождения разрешающего столбца
int findPivotColumn(const std::vector<std::vector<double>>& table, const std::vector<int>& basis) {
	int pivotCol = -1;
	double maxdelta = 0;
	for (int p = 0; p < N * N; p++) {
		double delta = -table[0][p];
		for (int i = 1; i <= basis.size(); i++) {
			delta += table[0][basis[i - 1]] * table[i][p];
		}
		if (delta > maxdelta) {
			maxdelta = delta;
			pivotCol = p;
		}
	}
	return pivotCol;
}

// Функция для нахождения разрешающей строки
int findPivotRow(const std::vector<std::vector<double>>& table, int pivotCol) {
	int pivotRow = -1;
	double minRatio = INF;
	for (int i = 1; i <= 2 * N; ++i) {
		if (table[i][pivotCol] > 0) {
			double ratio = table[i].back() / table[i][pivotCol];
			if (ratio < minRatio) {
				minRatio = ratio;
				pivotRow = i;
			}
		}
	}
	return pivotRow;
}

// Функция для выполнения симплекс-поворота
void pivot(std::vector<std::vector<double>>& table, std::vector<int>& basis, int pivotRow, int pivotCol) {
	double pivotValue = table[pivotRow][pivotCol];

	// Делим строку разрешающего элемента на сам элемент
	for (double& value : table[pivotRow]) {
		value /= pivotValue;
	}

	// Обновляем остальные строки
	for (int i = 1; i <= 2 * N; ++i) {
		if (i != pivotRow) {
			double factor = table[i][pivotCol];
			for (int j = 0; j < table[0].size(); ++j) {
				table[i][j] -= factor * table[pivotRow][j];
			}
		}
	}

	// Обновляем базисные переменные
	basis[pivotRow - 1] = pivotCol;
}

// Основная функция для решения задачи о назначениях с использованием симплекс-метода
void solveAssignmentProblem(std::vector<std::vector<double>>& cost) {
	// Инициализация симплекс-таблицы и базисных переменных
	std::vector<std::vector<double>> table(2 * N + 1, std::vector<double>(N * N + 1, 0));
	std::vector<int> basis(2 * N, -1);

	// Заполняем строку целевой функции
	for (int j = 0; j < N * N; ++j) {
		table[0][j] = cost[j / N][j % N];
	}

	// Заполняем таблицу ограничений
	for (int i = 1; i <= N; ++i) {
		for (int j = 0; j < N; ++j) {
			table[i][(i - 1) * N + j] = 1;
		}
		table[i][N * N] = 1;
	}
	for (int i = N + 1; i <= 2 * N; ++i) {
		for (int j = 0; j < N * N; ++j) {
			table[i][j] = ((j + 1) % N == (i - N) % N) ? 1 : 0;
		}
		table[i][N * N] = 1;
	}
	printTable(table, basis);
	// Заполнение массива basis
	for (int i = 0; i < N; ++i) {
		basis[i] = i * N;
	}
	int currentIndex = N;
	for (int i = 1; i <= N * N && currentIndex < 2 * N; ++i) {
		if (std::find(basis.begin(), basis.begin() + N, i) == basis.begin() + N) {
			basis[currentIndex++] = i;
		}
	}

	std::cout << "Начальный базис построен" << std::endl;
	printTable(table, basis);

	// Симплекс-метод
	int it=1;
	while (true) {


		int pivotCol = findPivotColumn(table, basis);
		if (pivotCol < 0) {
			break;
		}
		int pivotRow = findPivotRow(table, pivotCol);
		if (pivotRow == -1) {
			std::cout << "Решение не существует (неограниченная задача)." << std::endl;
			return;
		}

		pivot(table, basis, pivotRow, pivotCol);
		std::cout << "Итерация " << it << std::endl;
		it++;
		printTable(table, basis);
	}

	// Вывод оптимального решения
	double totalSum = 0;
	std::cout << "Оптимальное решение найдено:" << std::endl;
	for (int i = 0; i < basis.size(); ++i) {
		if (table[i + 1][N * N] != 0) {
			int col = basis[i] % N;
			int row = basis[i] / N;
			totalSum += cost[row][col];  // Добавляем соответствующее значение в сумму
			std::cout << "Работник " << row + 1 << " назначен на задачу " << col + 1 << std::endl;
		}
	}
	std::cout << "Сумма значений, соответствующих номерам в базисе: " << totalSum << std::endl;
}

int main() {
	// Ввод размерности
	std::cout << "Введите размер задачи N: ";
	std::cin >> N;

	// Инициализация таблицы стоимости
	std::vector<std::vector<double>> cost(N, std::vector<double>(N));

	std::cout << "Введите элементы таблицы стоимости:" << std::endl;
	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < N; ++j) {
			std::cin >> cost[i][j];
		}
	}

	solveAssignmentProblem(cost);

	return 0;
}
